### PR TITLE
[RISCV][ISEL] Lowering to load-acquire/store-release for RISCV Zalasr

### PR DESCRIPTION
**Pull request by mehnadnerd**
Lowering to load-acquire/store-release for RISCV Zalasr.

Currently uses the psABI lowerings for WMO load-acquire/store-release (which are identical to A.7). These are incompatable with the A.6 lowerings currently used by LLVM. This should be OK for now since Zalasr is behind the enable experimental extensions flag, but needs to be fixed before it is removed from that.

For TSO, it uses the standard Ztso mappings except for lowering seq_cst loads/store to load-acquire/store-release, I had Andrea review that.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td [hunk 1]
```diff
@@ -105,22 +105,66 @@ defm AMOMAXU_D  : AMO_rr_aq_rl<0b11100, 0b011, "amomaxu.d">,
...
+}
+
 // Atomic load/store are available under both +a and +force-atomics.
 // Fences will be inserted for atomic load/stores according to the logic in
 // RISCVTargetLowering::{emitLeadingFence,emitTrailingFence}.
 let Predicates = [HasAtomicLdSt] in {
-  def : LdPat<atomic_load_8,  LB>;
-  def : LdPat<atomic_load_16, LH>;
-  def : LdPat<atomic_load_32, LW>;
+  def : LdPat<relaxed_load<atomic_load_8>,  LB>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoA.td:156]: Maybe we need a comment to describe why we are using `relaxed` versions here (we have insert fences before).

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 2]
```diff
@@ -57,3 +57,47 @@ let Predicates = [HasStdExtZalasr, IsRV64] in {
...
+//===----------------------------------------------------------------------===//
+// Pseudo-instructions and codegen patterns
+//===----------------------------------------------------------------------===//
+
+class PatLAQ<SDPatternOperator OpNode, RVInst Inst, ValueType vt = XLenVT>
+    : Pat<(vt (OpNode (vt GPRMemZeroOffset:$rs1))), (Inst GPRMemZeroOffset:$rs1)>;
+
+class PatSRL<SDPatternOperator OpNode, RVInst Inst, ValueType vt = XLenVT>
+    : Pat<(OpNode (vt GPR:$rs2), (vt GPRMemZeroOffset:$rs1)),
+          (Inst GPRMemZeroOffset:$rs1, GPR:$rs2)>; // n.b. this switches order of arguments
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:70]: Move these notes above the definitions?

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 3]
```diff
@@ -57,3 +57,47 @@ let Predicates = [HasStdExtZalasr, IsRV64] in {
...
+                                                   // to deal with the fact that SRL has addr, data
+                                                   // while atomic_store has data, addr
+                              
+
+let Predicates = [HasStdExtZalasr] in {
+  def : PatLAQ<acquiring_load<atomic_load_8>, LB_AQ>;
+  def : PatLAQ<seq_cst_load<atomic_load_8>, LB_AQ>; // the sequentially consistent loads use
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:77]: Ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 4]
```diff
@@ -57,3 +57,47 @@ let Predicates = [HasStdExtZalasr, IsRV64] in {
...
+                                                    // .aq instead of .aqrl to match the psABI/A.7
+
+  def : PatLAQ<acquiring_load<atomic_load_16>, LH_AQ>;
+  def : PatLAQ<seq_cst_load<atomic_load_16>, LH_AQ>;
+
+  def : PatLAQ<acquiring_load<atomic_load_32>, LW_AQ>;
+  def : PatLAQ<seq_cst_load<atomic_load_32>, LW_AQ>;
+
+  def : PatSRL<releasing_store<atomic_store_8>, SB_RL>;
+  def : PatSRL<seq_cst_store<atomic_store_8>, SB_RL>; // the sequentially consistent stores use
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:87]: Ditto.

File: llvm/test/CodeGen/RISCV/atomic-load-store.ll [hunk 5]
```diff
@@ -5,13 +5,20 @@
 ; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-WMO %s
 ; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-ztso -verify-machineinstrs < %s \
 ; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-TSO %s
+; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-ZALASR,RV32IA-ZALASR-WMO %s
+; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr,+experimental-ztso -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-ZALASR,RV32IA-ZALASR-TSO %s
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/atomic-load-store.ll:11]: There is no difference between `RV32IA-ZALASR-TSO` and `RV32IA-TSO` (`RV64IA-ZALASR-TSO` and `RV64IA-TSO` also) currently, can we merge these CHECKs?

**Review by wangpc-pp — COMMENTED** [review]
These formatting changes don't belong to this PR, please remove them as they cause a lot of noises.

File: llvm/lib/Target/RISCV/RISCVISelLowering.h [hunk 6]
```diff
@@ -491,8 +491,8 @@ class RISCVTargetLowering : public TargetLowering {
       SDValue X, ConstantSDNode *XC, ConstantSDNode *CC, SDValue Y,
       unsigned OldShiftOpcode, unsigned NewShiftOpcode,
       SelectionDAG &DAG) const override;
-  /// Return true if the (vector) instruction I will be lowered to an instruction
-  /// with a scalar splat operand for the given Operand number.
+  /// Return true if the (vector) instruction I will be lowered to an
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.h:494]: Don't reformat unrelated code

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td [hunk 7]
```diff
@@ -105,22 +105,66 @@ defm AMOMAXU_D  : AMO_rr_aq_rl<0b11100, 0b011, "amomaxu.d">,
 // Pseudo-instructions and codegen patterns
 //===----------------------------------------------------------------------===//
 
+// An atomic load operation that does not need either acquire or release
+// semantics.
+class relaxed_load<PatFrags base>
+  : PatFrag<(ops node:$ptr), (base node:$ptr)> {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoA.td:111]: I think we have a preference 4 spaces before ':', but we're not consistent about it.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 8]
```diff
@@ -57,3 +57,47 @@ let Predicates = [HasStdExtZalasr, IsRV64] in {
...
+                                                      // .rl instead of .aqrl to match the psABI/A.7
+
+  def : PatSRL<releasing_store<atomic_store_16>, SH_RL>;
+  def : PatSRL<seq_cst_store<atomic_store_16>, SH_RL>;
+
+  def : PatSRL<releasing_store<atomic_store_32>, SW_RL>;
+  def : PatSRL<seq_cst_store<atomic_store_32>, SW_RL>;
+} // Predicates HasStdExtZalasr
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:95]: Predicates HasStdExtZalasr -> Predicates = [HasStdExtZalasr]

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 9]
```diff
@@ -57,3 +57,47 @@ let Predicates = [HasStdExtZalasr, IsRV64] in {
...
+
+let Predicates = [HasStdExtZalasr, IsRV64] in {
+  def : PatLAQ<acquiring_load<atomic_load_64>, LD_AQ>;
+  def : PatLAQ<seq_cst_load<atomic_load_64>, LD_AQ>;
+
+  def : PatSRL<releasing_store<atomic_store_64>, SD_RL>;
+  def : PatSRL<seq_cst_store<atomic_store_64>, SD_RL>;
+} // Predicates HasStdExtZalasr, IsRV64
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:103]: Predicates HasStdExtZalasr, IsRV64 => Predicates = [HasStdExtZalasr, IsRV64]

File: llvm/test/CodeGen/RISCV/atomic-load-store.ll (hunk 5 above)
Comment (mehnadnerd) [reply llvm/test/CodeGen/RISCV/atomic-load-store.ll:11]: There is a difference between ZALASR-TSO and TSO: the lowering of seq_cst loads and stores differs between them.
Between ZALASR-TSO and ZALASR-WMO, they differ in the lowering of acq loads and rel stores.
Since the RV32IA-ZALASR prefix takes precedence when they match, for seq_cst loads/stores they are listed as RV32AA-ZALASR loads/stores.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td [hunk 10]
```diff
@@ -105,22 +105,66 @@ defm AMOMAXU_D  : AMO_rr_aq_rl<0b11100, 0b011, "amomaxu.d">,
...
+  let IsAtomic = 1;
```
[the remark this answers is not in the document]
Comment (mehnadnerd) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:112]: I did it this way to match the AArch64 lowerings (llvm/lib/Target/AArch64/AArch64InstrAtomics.td), which this is adapted from.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td (hunk 1 above)
Comment (mehnadnerd) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:156]: The simple answer is that it lowers incorrectly (It uses plain loads and stores instead of load-acquire and store-release) if we don't use that pattern.
I am not entirely sure why that is the case, I believe that the default for atomic_load is to treat it as seq_cst.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td (hunk 10 above)
[the remark this answers is not in the document]
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:112]: Yes, I know it (I referred to AArhc64 too). :-)
I meant that we can reduce some lines via using above approaches.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td (hunk 1 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:156]: The reason is because we have inserted fences in AtomicExpandPass and this pass changes ordering to `AtomicOrdering::Monotonic` (or we can say `relaxed`). I meant we can add a comment line here to make this behavior clear.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td [hunk 11]
```diff
@@ -105,22 +105,65 @@ defm AMOMAXU_D  : AMO_rr_aq_rl<0b11100, 0b011, "amomaxu.d">,
...
+class seq_cst_store<PatFrag base>
+    : PatFrag<(ops node:$val, node:$ptr), (base node:$val, node:$ptr)> {
+  let IsAtomicOrderingSequentiallyConsistent = 1;
+}
+} // IsAtomic = 1
+
 // Atomic load/store are available under both +a and +force-atomics.
 // Fences will be inserted for atomic load/stores according to the logic in
 // RISCVTargetLowering::{emitLeadingFence,emitTrailingFence}.
+// The normal loads/store are relaxed (unordered) loads that don't have any
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoA.td:151]: `are relaxed (unordered) loads` -> `are relaxed (unordered) ones`, or just remove it?

File: llvm/test/CodeGen/RISCV/atomic-load-store.ll (hunk 5 above)
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/atomic-load-store.ll:11]: Thanks for clasifying! I think we can reduce some CHECKs line via adding `RV32IA-TSO` (and also for RV64)  to `-check-prefixes`.
```shell
; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr,+experimental-ztso -verify-machineinstrs < %s \
; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-ZALASR,RV32IA-TSO,RV32IA-ZALASR-TSO %s
```

File: llvm/test/CodeGen/RISCV/atomic-load-store.ll [hunk 12]
```diff
@@ -5,13 +5,20 @@
 ; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-WMO %s
 ; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-ztso -verify-machineinstrs < %s \
 ; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-TSO %s
+; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr -verify-machineinstrs < %s \
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/atomic-load-store.ll:8]: Sorry, I may have misled you. The CHECKs looks weird now.
Here is my example:
```diff
; RUN: llc -mtriple=riscv32 -mattr=+a -verify-machineinstrs < %s \                                                                                                                                                                        
 ; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-WMO %s                                                                                                                                                                                 
 ; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-ztso -verify-machineinstrs < %s \                                                                                                                                                     
-; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-TSO %s                                                                                                                                                                                 
+; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-TSO,RV32IA-NOZALASR-TSO %s                                                                                                                                                             
+; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr -verify-machineinstrs < %s \                                                                                                                                                   
+; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-ZALASR-WMO %s                                                                                                                                                                          
+; RUN: llc -mtriple=riscv32 -mattr=+a,+experimental-zalasr,+experimental-ztso -verify-machineinstrs < %s \                                                                                                                                
+; RUN:   | FileCheck -check-prefixes=RV32IA,RV32IA-TSO,RV32IA-ZALASR-TSO %s                                                                                                                                                               
 ; RUN: llc -mtriple=riscv64 -verify-machineinstrs < %s \                                                                                                                                                                                  
 ; RUN:   | FileCheck -check-prefix=RV64I %s
 ; RUN: llc -mtriple=riscv64 -mattr=+a -verify-machineinstrs < %s \
 ; RUN:   | FileCheck -check-prefixes=RV64IA,RV64IA-WMO %s
 ; RUN: llc -mtriple=riscv64 -mattr=+a,+experimental-ztso -verify-machineinstrs < %s \
-; RUN:   | FileCheck -check-prefixes=RV64IA,RV64IA-TSO %s
-
+; RUN:   | FileCheck -check-prefixes=RV64IA,RV64IA-TSO,RV64IA-NOZALASR-TSO %s
+; RUN: llc -mtriple=riscv64 -mattr=+a,+experimental-zalasr -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefixes=RV64IA,RV64IA-ZALASR-WMO %s
+; RUN: llc -mtriple=riscv64 -mattr=+a,+experimental-zalasr,+experimental-ztso -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefixes=RV64IA,RV64IA-TSO,RV64IA-ZALASR-TSO %s
```
Since there are differences (because of the difference of lowering `seq_cst`) between `RV32IA-TSO` and `RV32IA-ZALASR-TSO`, we need to add a prefix to the RUN line that is TSO without `Zalasr` to generate correct CHECKs. I name them `RV32IA-NOZALASR-TSO` and `RV64IA-NOZALASR-TSO`. I don't know if they are good names, maybe `RV32IA-STANDARD-TSO`, `RV32IA-NORMAL-TSO`, ...?

**Review by wangpc-pp — APPROVED** [review]
LGTM.
I don't have any more comment if the CHECKs are generated correctly. Thanks for your work!

**Comment by mehnadnerd** [comment]
I've updated it and it is now ARC approved https://lists.riscv.org/g/tech-unprivileged/message/966. Does anything else need to be done to merge it?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 13]
```diff
@@ -21926,6 +21926,35 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
   return isCtpopFast(VT) ? 0 : 1;
 }
 
+bool RISCVTargetLowering::shouldInsertFencesForAtomic(
+    const Instruction *I) const {
+  if (Subtarget.hasStdExtZalasr()) {
+    if (Subtarget.hasStdExtZtso()) {
+      // Zalasr + TSO means that atomic_load_acquire and atomic_store_release
+      //  should be lowered to plain load/store. The easiest way to do this is
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21934]: Only one space after `//`

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 14]
```diff
@@ -21926,6 +21926,35 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
...
+      //  to say we should insert fences for them, and the fence insertion code
+      //  will just not insert any fences
+      auto LI = dyn_cast<LoadInst>(I);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21937]: `auto *LI`. Dont' hide pointerness.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 15]
```diff
@@ -21926,6 +21926,35 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
...
+      auto SI = dyn_cast<StoreInst>(I);
+      if ((LI &&
+           (LI->getOrdering() == AtomicOrdering::SequentiallyConsistent)) ||
+          (SI &&
+           (SI->getOrdering() == AtomicOrdering::SequentiallyConsistent))) {
+        // Here, this is a load or store which is seq_cst, and needs a .aq or
+        //  .rl therefore we shouldn't try to insert fences
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21944]: Extra space

**Comment by mehnadnerd** [comment]
Should be fixed, I also made sure memory model is compatible, including cross-compatibility with tso

**Review by topperc — APPROVED** [review]
LGTM
